x-pack/plugin/otel-data/src/main/java/org/elasticsearch/xpack/oteldata/otlp/OTLPMetricsTransportAction.java

```diff
@@ -62,7 +80,131 @@ public OTLPMetricsTransportAction(
+                    .setRequireDataStream(true)
+                    .source(xContentBuilder)
+                    // TODO uncomment after https://github.com/elastic/elasticsearch/pull/132566 has been merged
+                    // .tsid(DataPointGroupTsidFunnel.forDataPointGroup(dataPointGroup).buildTsid())
+                    .setDynamicTemplates(dynamicTemplates)
+            );
+        }
+    }
+
+    private static void handleSuccess(ActionListener<MetricsResponse> listener) {
+        listener.onResponse(new MetricsResponse(RestStatus.OK, ExportMetricsServiceResponse.newBuilder().build()));
+    }
+
+    private static void handleEmptyRequest(ActionListener<MetricsResponse> listener) {
+        // If the server receives an empty request
+        // (a request that does not carry any telemetry data)
+        // the server SHOULD respond with success.
+        // https://opentelemetry.io/docs/specs/otlp/#full-success-1
+        handleSuccess(listener);
+    }
+
+    private static void handleEmptyBulk(ActionListener<MetricsResponse> listener, DataPointGroupingContext context) {
+        // If the processing of the request fails because the request contains data that cannot be decoded
+        // or is otherwise invalid and such failure is permanent,
+        // then the server MUST respond with HTTP 400 Bad Request.
+        // https://opentelemetry.io/docs/specs/otlp/#bad-data
+        listener.onResponse(
+            new MetricsResponse(
+                RestStatus.BAD_REQUEST,
+                responseWithRejectedDataPoints(context.totalDataPoints(), context.getIgnoredDataPointsMessage())
+            )
+        );
+    }
+
+    private static void handlePartialSuccess(
+        BulkResponse bulkItemResponses,
+        DataPointGroupingContext context,
+        ActionListener<MetricsResponse> listener
+    ) {
+        // If the request is only partially accepted
+        // (i.e. when the server accepts only parts of the data and rejects the rest),
+        // the server MUST respond with HTTP 200 OK.
+        // https://opentelemetry.io/docs/specs/otlp/#partial-success-1
+        RestStatus status = RestStatus.OK;
+        int failures = 0;
+        for (BulkItemResponse bulkItemResponse : bulkItemResponses.getItems()) {
+            failures += bulkItemResponse.isFailed() ? 1 : 0;
+            if (bulkItemResponse.isFailed() && bulkItemResponse.getFailure().getStatus() == RestStatus.TOO_MANY_REQUESTS) {
+                // If the server receives more requests than the client is allowed or the server is overloaded,
+                // the server SHOULD respond with HTTP 429 Too Many Requests or HTTP 503 Service Unavailable
+                // and MAY include “Retry-After” header with a recommended time interval in seconds to wait before retrying.
+                // https://opentelemetry.io/docs/specs/otlp/#otlphttp-throttling
+                status = RestStatus.TOO_MANY_REQUESTS;
+            }
```

Comment:
This makes it so that we return a 429 if there's at least one bulk response with a 429.
I think that's the best we can do given that OTLP doesn't have a per item response code.
If there's a mix of successes, failures, and 429s, the client will re-try the whole batch. That seems appropriate given that we have de-duplication in TSDB.

I've also discussed this with @axw in the context of the managed OTLP endpoint.